File: common/src/main/java/top/wcpe/wcpelib/common/mybatis/Mybatis.java
package top.wcpe.wcpelib.common.mybatis;

import org.apache.ibatis.datasource.pooled.PooledDataSource;
import org.apache.ibatis.mapping.Environment;
import org.apache.ibatis.session.Configuration;
import org.apache.ibatis.session.SqlSessionFactory;
import org.apache.ibatis.session.SqlSessionFactoryBuilder;
import org.apache.ibatis.transaction.jdbc.JdbcTransactionFactory;
import lombok.Getter;

public class Mybatis {
    @Getter
    private final String databaseName;

    public Mybatis(String url, int port, String database, String user, String password) {
        this(url, port, database, "useSSL=false&characterEncoding=UTF-8&serverTimezone=UTC", user, password);
    }

    public Mybatis(String url, int port, String database, String parameter, String user, String password) {
        this.databaseName = database;
        PooledDataSource pooledDataSource = new PooledDataSource("com.mysql.jdbc.Driver",
                "jdbc:mysql://" + url + ":" + port + "/" + database + "?" + parameter, user,
                password);
        pooledDataSource.setPoolMaximumActiveConnections(65535);
        pooledDataSource.setPoolMaximumIdleConnections(0);
        pooledDataSource.setPoolTimeToWait(10000);
        pooledDataSource.setPoolPingEnabled(true);
        pooledDataSource.setPoolPingQuery("SELECT 1");
        Environment environment = new Environment("development", new JdbcTransactionFactory(), pooledDataSource);
        Configuration configuration = new Configuration(environment);
        this.sqlSessionFactory = new SqlSessionFactoryBuilder().build(configuration);
    }

    @Getter
    private SqlSessionFactory sqlSessionFactory;


    public void addMapper(Class... classes) {
        if (sqlSessionFactory != null)
            for (Class clazz : classes) {
                sqlSessionFactory.getConfiguration().addMapper(clazz);
            }
    }

    public void addTypeHandler(TypeHandlerRegistry... typeHandlerRegistriess) {
        if (sqlSessionFactory != null)
            for (TypeHandlerRegistry typeHandlerRegistries : typeHandlerRegistriess) {
                sqlSessionFactory.getConfiguration().getTypeHandlerRegistry().register(typeHandlerRegistries.getJavaTypeClass(), typeHandlerRegistries.getJdbcType(), typeHandlerRegistries.getTypeHandlerClass());
            }
    }


}

File: common/src/main/kotlin/top/wcpe/wcpelib/common/mybatis/Mybatis.kt
package top.wcpe.wcpelib.common.mybatis

import org.apache.ibatis.session.SqlSession
import java.util.function.Consumer
import java.util.function.Function


/**
 * 由 WCPE 在 2022/1/3 21:47 创建
 *
 * Created by WCPE on 2022/1/3 21:47
 *
 * GitHub  : https://github.com/wcpe
 * QQ      : 1837019522
 * @author : WCPE
 * @since  : v1.0.7-alpha-dev-1
 */
object Mybatis {

    fun init(mybatisInstance: MybatisInstance): Mybatis {
        this.mybatisInstance = mybatisInstance
        this.database = mybatisInstance.getDatabaseName()
        return this
    }

    @JvmStatic
    lateinit var mybatisInstance: MybatisInstance

    lateinit var database: String


    fun getDatabaseName(): String {
        return mybatisInstance.getDatabaseName()
    }

    val sqlSessionFactory by lazy {
        mybatisInstance.sqlSessionFactory
    }

    val dataSource by lazy {
        mybatisInstance.dataSource
    }


    fun useSession(callBack: Consumer<SqlSession>) {
        mybatisInstance.useSession(callBack)
    }

    fun <R> useSessionForResult(callBack: Function<SqlSession, R>): R {
        return mybatisInstance.useSessionForResult(callBack)
    }

    fun addMapper(vararg classes: Class<*>?) {
        mybatisInstance.addMapper(*classes)
    }

    fun removeMapper(vararg classes: Class<*>?) {
        mybatisInstance.removeMapper(*classes)
    }
}

